### ISSUE TITLE
Improve `locktime` API 

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3334]
It was uncovered in #3290 that our `locktime` API is still somewhat lacking. To close this issue we need to:

- Implement suggestions in this post: https://github.com/rust-bitcoin/rust-bitcoin/pull/3290#issuecomment-2334971101
- Add docs as suggested by this post: https://github.com/rust-bitcoin/rust-bitcoin/pull/3290#issuecomment-2336434893
- Close #3319 by considering this post: https://github.com/rust-bitcoin/rust-bitcoin/pull/3319#issuecomment-2334961647

Warning, this issue is curly AF as is everything to do with Bitcoin locktimes. Please do not attempt this unless you are reasonably experience in API design and are willing to do a whole ton of iterations. 

Note also that reviewing the resulting PR will require a lot of attention and as such will be a drain on review resources. We should work out what the priority is on this issue before we attack it.

**Comment by jaoleal** [comment]
Please, lets reunite the bullet points of improvement here. 

Ill refrain the old ones that was in #3290. Note that Ill mention `Sequence` as actually being the `Relative Locktime` at all since itll cover 2/3 of its functions and because all the initial information and checks about `Relative Locktime`(indirectly about the transaction too) are extracted from the `Sequence` field. 

* In [bitcoin](https://docs.rs/bitcoin/0.32.2/bitcoin/index.html)::[relative](https://docs.rs/bitcoin/0.32.2/bitcoin/relative/index.html)::[LockTime](https://docs.rs/bitcoin/0.32.2/bitcoin/relative/enum.LockTime.html#) theres a wrong link in the docs while referring the bips. The link for the Bip68 is actually pointing to bip65.

* The check to see if the `Sequence` is valid is wrong for imitating the way to validate [absolute::LockTime](https://docs.rs/bitcoin/0.32.2/bitcoin/absolute/enum.LockTime.html#). Superficially, the correct way to validate sequence, if not only RBF and relative  locktime *feature* is disabled, is to take the Time/Height of the UTXO that the input is spending and adding the value extracted from the last 16 bits that is in sequence(if its a time notion value you have to multiply it by 512 since the 16bit value its a measure of 512 seconds intervals) and checking it against the updated values of the chain. If its the Height, check if is *less or equal* to the best block height. If its Time, check if its *less or equal* to the chain *Median Time Past*.

References for implementation!
* [Bip68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki)
* [Relative Locktime and sequence explanation in learnmeabitcoin](https://learnmeabitcoin.com/technical/transaction/input/sequence/)
* [ Median Time Past mention in learnmeabitcoin](https://learnmeabitcoin.com/technical/block/time/).

**Comment by tcharding** [comment]
> In [bitcoin](https://docs.rs/bitcoin/0.32.2/bitcoin/index.html)::[relative](https://docs.rs/bitcoin/0.32.2/bitcoin/relative/index.html)::[LockTime](https://docs.rs/bitcoin/0.32.2/bitcoin/relative/enum.LockTime.html#) theres a wrong link in the docs while referring the bips. The link for the Bip68 is actually pointing to bip65.

Oh bother, we should just fix that straight away.

**Comment by jaoleal** [comment]
For anyone thats seeking a way to correctly validate `Relative Locktime` with the 0.32.0 release from rust-bitcoin (the latest rn)

Thats how im doing in Floresta

```Rust
 if input.sequence.is_relative_lock_time() {
            //validate lock time
            let prevout = out_map.get(&input.previous_output).unwrap();
            if input.sequence.is_height_locked() {
                let commited_height = prevout.get_height();
                //to retrieve the span contained in the sequence we have to just get the u32 value that the sequence contains.
                let height_span = input.sequence.0;

                // if the commited height + the span is greater than the current height, the transaction is invalid.
                if commited_height + height_span > height {
                    return Err(BlockValidationErrors::BadRelativeLockTime);
                }
            }
            if input.sequence.is_time_locked() {
                let commited_time = prevout.get_time();
                // here we have to shift the sequence 16 bits to the left and 16 to the right get the time lock without the flag messing with us.
                let time_lock = ((input.sequence.0 << 16) >> 16) * 512 as u32;

                if commited_time + time_lock > mtp {
                    return Err(BlockValidationErrors::BadRelativeLockTime);
                }
            }
        }
       
        }
```

**Comment by apoelstra** [comment]
> * The check to see if the `Sequence` is valid is wrong

Can we also fix this without touching any other part of the API? This sounds like it's worth backporting too.

**Comment by tcharding** [comment]
> The check to see if the Sequence is valid is wrong for imitating the way to validate [absolute::LockTime](https://docs.rs/bitcoin/0.32.2/bitcoin/absolute/enum.LockTime.html#).

I don't believe the claim in this paragraph is true, I believe the current API correctly allows for this. I do admit that it may be hard to understand how to use it. I do not want to arguing the position right here, I'm just throwing it up so no one blindly follows the comment above (I could of course be wrong).

**Comment by jaoleal** [comment]
>I don't believe the claim in this paragraph is true, I believe the current API correctly allows for this. I do admit that it may be hard to understand how to use it. I do not want to arguing the position right here, I'm just throwing it up so no one blindly follows the comment above (I could of course be wrong).

I took some time to understand what you mean. And you're right but still wrong.
Let me explain with rust code: 
```Rust

// tobin's idea (i guess)
let heightspan_or_distance: u16 = Sequence.extract_value() // pretend that the extracted value is like 100.

let prevout_commitment_height: u32 = 200_000

let best_block = 200_101

let time_lock = Sequence::from( heightspan_or_distance + prevout_commitment_height ).to_locktime()

timelock.is_satisfied_by_height(best_block);

```

This hipothetical rust code will not work at all. Because the maximum value that we can store in the Sequence/Timelock is a u16.
this was right back there in 2010 when all this values could be stored inside two u16 variables, but not today.
This is reforced by the @apoelstra [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/3290#discussion_r1747833415) when we got confused because `Time` that i was talking was one of the two timestamps that, together with the value in the Sequence, we need to check against. and apoelstra is thinking about a u16 intervals(i guess).

That said, the addition method before building the sequence could not be possible because we dont have enough range. 

Please, remember that the actual API makes you build a 

```Rust
// In https://docs.rs/bitcoin-units/0.1.1/src/bitcoin_units/locktime/relative.rs.html#13
pub struct Height(u16); // Same for Time. 
```
as input to make the comparisons.

Event that, we change the values to be a u32. we could miss some range in the future. (Likely we will be dead in this future so we will be lefting a nice gift for the ones in 83730 A.C. maintaining this code ).

Jokes and suspicious math aside. For the addition method to work, we need atleast to change the internal value in the variants that Locktime can be to a u32. And, in my opinion, its not that intuitive but thats on you guys.

**Comment by Kixunil** [comment]
@jaoleal I'm confused what you're talking about since changing the type would be a hard fork. There never was one apart from the db-lock-count-relaxing one and the next one is only due until Sun Feb  7 07:28:15 AM CET 2106, so I don't think people are too worried about it.

**Comment by tcharding** [comment]
> let time_lock = Sequence::from( heightspan_or_distance + prevout_commitment_height ).to_locktime()

We don't want to add these two values together in the sequence number, that would be putting an absolute height into the sequence number where as it should be a relative height i.e., onyl the `heightspan_or_distance` variable.

From BIP-68:

> More generally, a relative block lock-time n can be included n blocks after the mining date of the output it is spending, or > any block thereafter.

**Comment by jaoleal** [comment]
So, can you please provide an usage example using the actual API to correctly validate Relative Locktime ? I really dont have a clue about what are we disagreeing.

**Comment by Davidson-Souza** [comment]
@jaoleal I think @tcharding is right, the code is correct. After reading the code and trying it out with real transactions, I could see where the main misunderstanding were. `Height` and `Time` are incredibly misleading names, they don't represent the actual height (or time) for that transaction, but the delta since this UTXO creation. You should compute those **before** calling the API. Your patch works when the user gives the actual times or heights for the creation and spending of a UTXO. Example:

Let A = A UTXO created at block 800_000, and T a transaction that got confirmed at block 800_100 spending A, with a relative locktime of 50.

For the current API, the user would do this (pseudocode):
```rust
let delta = 800_100 - 800_000;
let heigth = bitcoin::relative::Height::from_height(delta); // notice that you can't create this from the actual height (u16 won't fit)
let is_satisfied = A.sequence.is_satisfied_by_height();  
assert!(is_satisfied);
````

Your patch does this:
```rust
let is_satisfied = A.sequence.is_satisfied_by_height(800_000, 800_100);
assert!(is_satisfied);
```

IMO, either way is fine. However, the naming here could be improved.

**Comment by jaoleal** [comment]
So, the PR to solve #3334 should be named  "Improve Locktime Docs and naming" rather than changing any code. 

Btw, if @Davidson-Souza  is right about the how were supposed to use this code. Theres a way to implement `is_satisfied_by(Time/Height)` only by giving the **delta** as a `u16` value...

**Comment by Kixunil** [comment]
Perhaps we should more strongly recommend the `relative::` prefix, especially for the relative locks (absolute are pretty much fine). But I still like the `MtpAndHeight` idea.

> Everything about Bitcoin locktimes is non-intuitive.

I personally don't find it that bad but maybe because I understand Bitcoin way too much. :D Actually, it's a good idea to explain this weirdness in a workshop I'm preparing...

**Comment by Davidson-Souza** [comment]
Thanks for the clarification @tcharding. I think the consensus here is that we need more documentation for the locktime API, perhaps a cookbook entry?

**Comment by tcharding** [comment]
A cookbook entry would be nice but this issue, once close, will add a whole new API. I'm not sure how much work its worth putting into documenting the current one further.

**Comment by jaoleal** [comment]
>A cookbook entry would be nice but this issue, once close, will add a whole new API. I'm not sure how much work its worth putting into documenting the current one further.
So are we remaking the Relative locktime api ? but without breaking it ? 

I could stick on this but i already have a Pr that i need to get done... (I have interest btw).

**Comment by apoelstra** [comment]
We won't break the existing API (I hope), but we are adding the new `MtpAndHeight` structure and associated methods to the locktime types, and that will change the "recommended usage" of the API for users who are trying to determine a transaction's validity.

**Comment by tcharding** [comment]
Note to dev, check https://github.com/rust-bitcoin/rust-bitcoin/pull/3364/files#r1764040105 in case it was useful.

**Comment by tcharding** [comment]
I've re-read this thread a couple of times and have no clue what to do about the issue. 

- I don't know where the `MtpAndHeight` idea referenced twice above comes from?
- I don't know exactly what docs to improve, if any, because I believe that everytime one tries to better describe locktimes they introduce more ambiguity. Happy to review other peoples attempts though


I'm inclined to think that this issue can just be closed.

**Comment by Kixunil** [comment]
```rust
pub struct MtpAndHeight {
    /*pub*/ time: Timestamp,
    /*pub*/ height: Height,
}

impl MtpAndHeight {
    from_recent_headers(current _height: Height, headers: &[impl Borrow<block::Header>; 11]) -> Self;
}
```

**Comment by apoelstra** [comment]
Yes, that is a start -- though I also have forgotten what associated methods we want for the locktime types, or how we want to use this to validate relative locktimes.

**Comment by sanket1729** [comment]
I ran into another issue with locktime API. 


```rust
    /// Adds instructions to push an absolute lock time onto the stack.
    pub fn push_lock_time(self, lock_time: absolute::LockTime) -> Builder {
        self.push_int(lock_time.to_consensus_u32().into())
    }

    /// Adds instructions to push a sequence number onto the stack.
    pub fn push_sequence(self, sequence: Sequence) -> Builder {
        self.push_int(sequence.to_consensus_u32().into())
    }
```

I think the second API should be to push `relative::LockTime`. As it stands, I think pushing `Sequence` is rather footgunny. We push the timelock values in script and compare those against `Sequence`. Directly pushing Sequence does not make much sense? 

Sorry if was discussed before, but I just ran into this in my own project.

**Comment by Kixunil** [comment]
@sanket1729 good point! We should definitely change it.

**Comment by apoelstra** [comment]
Closing this issue. It's very hard to follow and discusses multiple indepedent problems.

Split into

* https://github.com/rust-bitcoin/rust-bitcoin/issues/4299 add `MtpAndHeight`
* https://github.com/rust-bitcoin/rust-bitcoin/issues/4300 sequence validity check is wrong
* https://github.com/rust-bitcoin/rust-bitcoin/issues/4301 script should push locktimes, not sequence numbers
